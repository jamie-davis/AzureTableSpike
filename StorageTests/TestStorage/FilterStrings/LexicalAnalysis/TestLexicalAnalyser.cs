using ApprovalTests;
using ApprovalTests.Reporters;
using NUnit.Framework;
using StorageTests.TestUtilities;
using TestConsole.OutputFormatting;
using TestConsoleLib;
using TestStorage.FilterStrings.LexicalAnalysis;

namespace StorageTests.TestStorage.FilterStrings.LexicalAnalysis
{
    [TestFixture]
    [UseReporter(typeof(CustomReporter))]
    public class TestLexicalAnalyser
    {
        [Test]
        public void LexicalAnalysisTests()
        {
            //Arrange
            var testStrings = new[]
            {
                "var eq 'literal'",
                "(var gt 'literal' and var2 lt 45) or (datevar gt datetime'2017-07-21T22:02:31Z')",
                "var eq guid'9E37E338-27B2-4F32-AFFA-DC74F017AF1D' or (var2 ge 45L and var3 ne 35.4)",
                string.Empty,
                null,
                "\t\tvar        eq\r\n'literal'",
            };

            var output = new Output();

            //Act
            foreach (var testString in testStrings)
            {
                var result = LexicalAnalyser.Analyse(testString);

                output.WrapLine($@"Analysis of ""{testString ?? "NULL"}"":");
                output.FormatTable(result.AsReport(rep => rep
                                                    .AddColumn(r => r.TokenType, cc => cc.LeftAlign())
                                                    .AddColumn(r => r.Text, cc => {}))
                );

                output.WriteLine();
                output.WriteLine();
            }

            //Assert
            Approvals.Verify(output.Report);
        }
    }
}
