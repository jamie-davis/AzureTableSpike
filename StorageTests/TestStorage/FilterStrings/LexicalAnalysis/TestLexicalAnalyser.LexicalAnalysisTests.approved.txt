Analysis of "var eq 'literal'":                                                                
Token Type    Text                                                                             
------------- -------                                                                          
Identifier    var                                                                              
CompEq        eq                                                                               
StringLiteral literal                                                                          
                                                                                               
                                                                                               
Analysis of "(var gt 'literal' and var2 lt 45) or (datevar gt datetime'2017-07-21T22:02:31Z')":
Token Type      Text                                                                           
--------------- --------------------                                                           
OpenParen       (                                                                              
Identifier      var                                                                            
CompGt          gt                                                                             
StringLiteral   literal                                                                        
LogicalAnd      and                                                                            
Identifier      var2                                                                           
CompLt          lt                                                                             
IntLiteral      45                                                                             
CloseParen      )                                                                              
LogicalOr       or                                                                             
OpenParen       (                                                                              
Identifier      datevar                                                                        
CompGt          gt                                                                             
DateTimeLiteral 2017-07-21T22:02:31Z                                                           
CloseParen      )                                                                              
                                                                                               
                                                                                               
Analysis of "var eq guid'9E37E338-27B2-4F32-AFFA-DC74F017AF1D' or (var2 ge 45L and var3 ne 35. 
4)":                                                                                           
Token Type    Text                                                                             
------------- ------------------------------------                                             
Identifier    var                                                                              
CompEq        eq                                                                               
GuidLiteral   9E37E338-27B2-4F32-AFFA-DC74F017AF1D                                             
LogicalOr     or                                                                               
OpenParen     (                                                                                
Identifier    var2                                                                             
CompGe        ge                                                                               
LongLiteral   45                                                                               
LogicalAnd    and                                                                              
Identifier    var3                                                                             
CompNe        ne                                                                               
DoubleLiteral 35.4                                                                             
CloseParen    )                                                                                
                                                                                               
                                                                                               
Analysis of "":                                                                                
Token Type   Text                                                                              
------------ ----                                                                              
                                                                                               
                                                                                               
Analysis of "NULL":                                                                            
Token Type   Text                                                                              
------------ ----                                                                              
                                                                                               
                                                                                               
Analysis of "        var        eq                                                             
'literal'":                                                                                    
Token Type    Text                                                                             
------------- -------                                                                          
Identifier    var                                                                              
CompEq        eq                                                                               
StringLiteral literal                                                                          
                                                                                               
                                                                                               
                                                                                               